snapcraft/internal/pluginhandler/__init__.py

```diff
@@ -965,20 +965,42 @@ def _handle_elf(self, snap_files: Sequence[str]) -> Set[str]:
         if not self._build_attributes.keep_execstack():
             clear_execstack(elf_files=elf_files)
 
-        if self._build_attributes.no_patchelf():
-            logger.warning(
-                "The primed files for part {!r} will not be verified for "
-                "correctness or patched: build-attributes: [no-patchelf] "
-                "is set.".format(self.name)
-            )
-        else:
-            part_patcher = PartPatcher(
-                elf_files=elf_files,
-                project=self._project,
-                snap_base_path=self._snap_base_path,
-                stage_packages=self._part_properties.get("stage-packages", []),
+        # ELF files in this part need to have their rpath and interpreter patched
+        # to use the in-snap version in the following scenario:
+        #
+        #   - `enable-patchelf` is a build attribute
+        #   OR
+        #     - The base is defined
+        #     AND
+        #       - The base is not one of the static bases
+        #     AND
+        #       - The snap uses classic confinement
+        #       OR
+        #         - libc has been staged (as opposed to being in the base snap)
+        patching_required = self._build_attributes.enable_patchelf() or (
```

Comment:
This logic feels like we have 3 groups:

- patching required (everything in the latter statement of this logical or)
- patching desired
- pathing required but ignored (no-patchelf)
- patching desired (enable-patchelf) but ignored (no-patchelf)

That last one should not be possible and yet considered in the if below.

I would change this to
```
if patching_required and no_pacthelf():
    logger.warning(...)
elif patching_required or enable_patchelf():
    patch(...)
# For precaution
elif enable_patchelf() and no_patchelf():
    raise RuntimeError(...)
```

Reply:
Good call, done.